자동차 경주 2단계

wishwisdom opened the pull request:
자동차 경주 2단계

step0와 섞인 것 삭제

Car 
 - name 추가
   .toString에 name을 추가
   .constructor의 parameter를 name을 받아옮

Main
 - driver들의 이름을 받아옮
 - 최종 우승자를 출력함
 
RacingGame
 - 입력 받은 이름을 잘라, 이름 갯수만큼 Car 객체를 만들
 - getMaxDistance 
   . 가장 멀리간 드라이버의 위치를 가져오는 메소드를 생성
 - findBestDriver
   . 가장 멀리 간 드라이버를 찾음
 - findBestDrivers
   . 가장 멀리 간 드라이버들을 찾음
 - getBestDriverNames
   . 가장 멀리간 드라이버의 이름을 찾음
   
 RacingInput
  - 출력 텍스트 변경
  
 RacingResult
  - 최종 우승자 출력해주는 부분 추가

shinhong-codelinkcorp committed:
rebase

javajigi commented on src/main/java/racing/Car.java:
```diff
@@ -1,17 +1,11 @@
 
 public class Car {
-    private static int RACING_NUM_DETERMINE = 1;
     private StringBuilder location;
```
이와 같이 관리하는 경우 너무 View와 종속적인 코드가 될 것 같아요.
View와 의존적이지 않도록 위치 값을 integer로 가지면 어떨까요?

javajigi commented on src/main/java/racing/Car.java:
```diff
@@ -25,6 +19,15 @@ public String getLocation(){
 
```
random 값에 따른 위치 변경 로직을 Car가 가지면 어떨까요?

car.move(int randomValue);

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -3,26 +3,25 @@
     private List<Car> racingCars;
     private Random random;
-    private int size;
+    private String[] driverNames;
```
driverNames 정보들은 이미 List<Car>에 있는데 굳이 인스턴스로 가지고 있을 필요가 있을까?

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -49,4 +48,32 @@ public void move(Car car, int num){
+    }
+
+    public void findBestDriver(int bestDistance, List<Car> bestDrivers, Car car) {
+        if(car.getDistance() == bestDistance){
```
상태를 가지는 객체의 데이터를 꺼내려(get)하지 말고 객체에 메시지를 보내라는 원칙에 따라 구현한다.

car.matchDistance(bestDistance)

javajigi reviewed (changes requested):
Car 객체의 위치 값 정보를 어떻게 관리하는 것이 좋을지 한번 고민해 좋겠어요.

shinhong-codelinkcorp committed:
Car
 - StringBuilder location -> int position
   .상태만 가지도록 변경
 - matchDistance 추가
   .객체의 변화는 스스로 변화하게 만듦

 RacingGame
  - readyRacing에서 값을 가짐
    .생성자에 이름을 넣어줄 경우, 경기시작 전에 객체르 생성하기 때문에,
     readyRacing에서 객체생성하도록 설계하여서 생성자에 driverName을 넣지 않음
  - for문을 stream으로 만들어 사용함.
